core/trino-main/src/main/java/io/trino/operator/join/DefaultPagesHash.java

```diff
@@ -175,6 +175,78 @@ public int getAddressIndex(int rightPosition, Page hashChannelsPage, long rawHas
         return -1;
     }
 
+    @Override
+    public int[] getAddressIndex(int[] positions, Page hashChannelsPage)
+    {
+        long[] hashes = new long[positions[positions.length - 1] + 1];
+        for (int i = 0; i < positions.length; i++) {
+            hashes[positions[i]] = pagesHashStrategy.hashRow(positions[i], hashChannelsPage);
+        }
+
+        return getAddressIndex(positions, hashChannelsPage, hashes);
+    }
+
+    @Override
+    public int[] getAddressIndex(int[] positions, Page hashChannelsPage, long[] rawHashes)
+    {
+        int positionCount = positions.length;
+        int[] hashPositions = new int[positionCount];
+
+        for (int i = 0; i < positionCount; i++) {
+            hashPositions[i] = getHashPosition(rawHashes[positions[i]], mask);
+        }
+
+        int[] found = new int[positionCount];
+        int foundCount = 0;
+        int[] result = new int[positionCount];
+        Arrays.fill(result, -1);
+        int[] foundKeys = new int[positionCount];
+
+        // Search for positions in the hash array. This is the most CPU-consuming part as
+        // it relies on random memory accesses
```

Comment:
`positionEqualsCurrentRowIgnoreNulls` can be actually way more expensive as depending on the number of channels it can have a lot more random memory accesses + it has to do the the `equals` logic

Reply:
Correct. Batching some methods from `PagesHashStrategy` might be a good follow-up